src/collection.jl

```diff
@@ -115,30 +172,107 @@ end
 
 
 """
-    filenames(df::DataFrame)
+    filenames(collection::DataFrame)
 
 Generator for filenames of entries in data frame.
 """
 function filenames end
 
 # generator for filenames specified by data frame (i.e. path of file, hdu etc.)
-@resumable function filenames(df::DataFrame)
-    for row in eachrow(df)
+@resumable function filenames(collection::DataFrame)
+    for row in eachrow(collection)
         @yield row.path
     end
 end
 
 
 """
-    images(df::DataFrame)
+    images(collection::DataFrame)
 
 Generator for `ImageHDU`s of entries in data frame.
 """
 function images end
 
 # generator for ImageHDU specified by data frame (i.e. path of file, hdu etc.)
-@resumable function images(df::DataFrame)
-    for row in eachrow(df)
+@resumable function images(collection::DataFrame)
+    for row in eachrow(collection)
         @yield FITS(row.path)[row.hdu]
     end
 end
+
+
+"""
+    images(f, collection::DataFrame; save = false, path = nothing, save_prefix = nothing, save_suffix = nothing, save_delim = "_", ext = r"fits(\\.tar\\.gz)?"i)
+
+Applies function `f` on all ImageHDUs present in data frame and saves it in FITS file.
+
+It returns an array of output values of function `f` applied on ImageHDUs. If `save = false`, the save functionality does not execute. File is saved at `path` specified by the user.
+Suffix and prefix can be added to filename of newly created files by modifying `save_suffix` and `save_prefix`, `save_delim` is used as delimiter.
+`ext` is the extension of files to be taken into consideration for applying function, by default it is set to `r"fits(\\.tar\\.gz)?"i`.
+"""
+function images(f, collection::DataFrame; save = false, path = nothing, save_prefix = nothing, save_suffix = nothing, save_delim = "_", ext = r"fits(\.tar\.gz)?"i, kwargs...)
+    image_iterator = images(collection; kwargs...)
+    locations = collection.path
+
+    processed_images = map(zip(locations, image_iterator)) do (location, output)
+        processed_image = f(output)
+        if save
+            save_path = generate_filename(location, path, save_prefix, save_suffix, save_delim, ext)
+            write_data(save_path, processed_image)
+        end
+        processed_image
+    end
+
+    return processed_images
+end
+
+"""
+    filenames(f, collection::DataFrame; save = false, path = nothing, save_prefix = nothing, save_suffix = nothing, save_delim = "_", ext = r"fits(\\.tar\\.gz)?"i)
+
+Applies function `f` on all file paths present in data frame and saves it in FITS file.
```

Comment:
very important! It does *not* always save it in a FITS file. You should write this docstring as "this iterates over the filenames of the collection applying function `f` at each step. ... ... ... The following parameters enable programmatically saving the returned value `f` using [`CCDReduction.write_fits`](@ref)"